src/MyParcelComApi.php

```diff
@@ -398,6 +417,53 @@ public function getServiceRatesForShipment(ShipmentInterface $shipment, $ttl = s
         return new ArrayCollection($matchingServiceRates);
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function resolveDynamicServiceRates($shipmentData, $dynamicServiceRate = null)
+    {
+        $data = ($shipmentData instanceof ShipmentInterface) ? $shipmentData->jsonSerialize() : $shipmentData;
+
+        if (!isset($data['relationships'])) {
+            $data['relationships'] = [];
+        }
+        if (!isset($data['relationships']['shop'])) {
+            $data['relationships']['shop'] = [
+                'data' => [
+                    'type' => ResourceInterface::TYPE_SHOP,
+                    'id'   => $this->getDefaultShop()->getId(),
+                ],
+            ];
+        }
+
+        if ($dynamicServiceRate) {
+            $data['relationships']['service'] = [
+                'data' => [
+                    'type' => ResourceInterface::TYPE_SERVICE,
+                    'id'   => $dynamicServiceRate->getService()->getId(),
+                ],
+            ];
+            $data['relationships']['contract'] = [
+                'data' => [
+                    'type' => ResourceInterface::TYPE_CONTRACT,
+                    'id'   => $dynamicServiceRate->getContract()->getId(),
+                ],
+            ];
+        }
+
+        if (!isset($data['relationships']['service'])) {
+            throw new InvalidResourceException('Cannot retrieve dynamic rates without a service');
+        }
+        if (!isset($data['relationships']['contract'])) {
+            throw new InvalidResourceException('Cannot retrieve dynamic rates without a contract');
+        }
```

Comment:
We have more requirements for this endpoint, like the sender and recipient addresses and weight. 
Would you happen to know how the SDK handles spec errors? It feels arbitrary to check service and contract, but not the other requirements here.

Reply:
I will 🔥 this because I have just added request validation on this endpoint in our API.